Add a quadrature phase tracker

jclsn opened the pull request:
Depends on https://github.com/mixxxdj/mixxx/pull/15194

The current pitch filter works under the assumption that zero crossings are never skipped. Unfortunately this happens quite frequently when frequencies are very high during e.g. scratching or backspins. Especially during backspins the pitch filter was being starved by always feeding a quarter rotation displacement, when in reality the displacement was a half rotation or three quarters even.

This commit fixes this issue by keeping track of the quadrant we are in. The number of the last quadrant we were in, is compared to the current number. Each quadrant corresponds to 90° and is quantized to timecode revolutions.

This fixes backspins and improves pitch robustness when the needle runs over a scratch or during audio buffer dropouts. 

I also lifted the `ZERO_THRESHOLD` for detecting zero crossings for the MK2s. This improved crossing detection.

jclsn committed:
xwax: Add a quadrature phase tracker

The current pitch filter works under the assumption that zero crossings
are never skipped. Unfortunately this happens quite frequently when
frequencies are very high during e.g. scratching or backspins. Especially
during backspins the pitch filter was being starved by always feeding a
quarter rotation displacement, when in reality the displacement was a half
rotation or three quarters even.

This commit fixes this issue by mapping the quadrants integers 1..4.
The last quadrant we were in, is compared to the current one and the
difference is calculated. A difference of 1 corresponds to a 90° phase angle.
This rotation angle is then quantized to timecode revolutions.

This fixes backspins and improves pitch robustness when the needle runs
over a scratch or during audio buffer dropouts.

fixup

Swiftb0y commented:
But it wouldn't be integers then.

Why? If you just translated the current version into a polar form, you would still operate on the floating point values $$0$$, $$\pi \over 2$$, $$\pi$$, $$3 \pi \over 4$$, so not integer compatible. By dividing that scale by $2 \pi$, you would then operate on $0$, $1$, $2$, $3$, so now only integers. And then in order to stay in the $[0;2\pi[$ domain (which is now the $[0; 4[$ domain), you'll just have to $\mod 4$ sometimes.

jclsn commented:
> But it wouldn't be integers then.
> 
> Why? If you just translated the current version into a polar form, you would still operate on the floating point values 0 , π 2 , π , 3 π 4 , so not integer compatible. By dividing that scale by 2 π , you would then operate on 0 , 1 , 2 , 3 , so now only integers. And then in order to stay in the [ 0 ; 2 π [ domain (which is now the [ 0 ; 4 [ domain), you'll just have to mod 4 sometimes.

I already manged to implement an integer-based phase tracker, but it doesn't work as expected yet.

Swiftb0y commented:
If you push, I may be able to spot the bug

jclsn committed:
xwax: Add ZERO_THRESHOLD to struct timecode_def

While working on the phase tracker I discovered that many zero crossings
are skipped. There were also too many direction changes detected for the
MK2s. I could improve this by lifting the threshold.

This commits also adds it to the timecode definition. It's not clear if
the default threshold works well for other timecodes then the Seratos.
With this structure thresholds can be changed per timecode.

We might also want to dynamically calibrate a threshold to the
RMS value of the carrier wave. This would probably yield better results
than using these fixed ones.

jclsn commented:
@Swiftb0y The solutions works with simple integer arithmetic now.

Swiftb0y commented on lib/xwax/timecoder.c:
```diff
@@ -430,7 +431,12 @@ void timecoder_init(struct timecoder *tc, struct timecode_def *def,
     tc->sample_rate = sample_rate;
     tc->zero_alpha = tc->dt / (ZERO_RC + tc->dt);
-    tc->threshold = ZERO_THRESHOLD;
+
+    if (tc->def->flags & TRAKTOR_MK2)
+        tc->threshold = ZERO_THRESHOLD_MK2;
+    else
+        tc->threshold = ZERO_THRESHOLD;
+
```
NIT: consider moving to `timecode_def` instead of adding yet more `tc->def->flags & TRAKTOR_MK2` (IMO) hacks?

Swiftb0y commented on lib/xwax/timecoder.c:
```diff
@@ -629,6 +639,59 @@ static void process_bitstream(struct timecoder *tc, signed int m)
+
+    } else if ((tc->forwards && diff == 3) || (!tc->forwards && diff == 1) ) { /* Check for a displacement of three quadrants */
+        debug("Skipped 2 quadrants\n");
+        return (3.0 / tc->def->resolution) / 4.0;
```
can you quickly explain where the $$n / \text{res} / m$$  formula and the values for `n`/`m` come from?

jclsn commented on lib/xwax/timecoder.c:
```diff
@@ -629,6 +639,59 @@ static void process_bitstream(struct timecoder *tc, signed int m)
+    tc->last_quadrant = tc->quadrant;
+    tc->direction_changed = direction_changed;
+
+    if (tc->primary.swapped && tc->primary.positive)
+        tc->quadrant = 1;
+    else if (tc->primary.swapped && !tc->primary.positive)
+        tc->quadrant = 3;
+    else if (tc->secondary.swapped && tc->secondary.positive)
+        tc->quadrant = 2;
+    else if (tc->secondary.swapped && !tc->secondary.positive)
+        tc->quadrant = 4;
+}
```
They should be in theory. There can always be errors though.

jclsn commented on lib/xwax/timecoder.c:
```diff
@@ -629,6 +639,59 @@ static void process_bitstream(struct timecoder *tc, signed int m)
+
+    if (tc->direction_changed) {
+        direction_change_counter++;
+        debug("Direction changed %d times\n", direction_change_counter);
```
We can also remove those prints. Unfortunately you can't do selective debug loggin in xwax and they kind of drown in the rest of the logs.

acolombier commented on lib/xwax/timecoder.c:
```diff
@@ -653,24 +732,42 @@ static void process_sample(struct timecoder *tc,
 	    dx = -dx;
-	pitch_dt_observation(&tc->pitch, dx);
+
+        if (tc->use_legacy_pitch_filter)
+            pitch_dt_observation(&tc->pitch, dx);
+        else
+            pitch_kalman_update(&tc->pitch_kalman, dx);
```
Is this condition block conditional to `!tc->forwards`? If yes, you need add curly bracket at line 764, otherwise, perhaps remove the indentation

ywwg replied:
yeah it looks like the 764 has a tab, which is rendered as 2 spaces, and 767 is 8 spaces.

jclsn replied:
Mark had his editor misconfigured once. He just recently fixed this after reading that I publicly stated in the MK2 PR that formatting is not consistent in xwax. 

I can can make a small formatting PR. Maybe even add a clang-format file to the xwax folder. Usually this is discouraged in large projects, but in this case it's fine I think.

ywwg replied:
Yeah unfortunately if we do a mass formatting fix it will make merging/diffing with Mark's code harder.  Many code formatters provide the option to only run formatting on changed lines, that might be the thing to do here.  Not blocking for this PR though!

jclsn replied:
I added a kernel-based clang-format file with minor changes and formatted the whole library.

acolombier reviewed (commented):
Didn't test, mostly code style feedback. Thanks for looking into this!

Swiftb0y commented on lib/xwax/timecoder.c:
```diff
@@ -615,6 +639,59 @@ static void process_bitstream(struct timecoder *tc, signed int m)
+
+
+/*
+ * Track the quadrature phase of the pitch counter in the complex plane
+ *
+ * Translates the zero crossings into the complex plane for a proper
+ * mathematica
```
NIT: outdated after the `unsigned` rework.

ywwg commented on lib/xwax/timecoder.c:
```diff
@@ -428,14 +431,35 @@ void timecoder_init(struct timecoder *tc, struct timecode_def *def,
     tc->zero_alpha = tc->dt / (ZERO_RC + tc->dt);
-    tc->threshold = ZERO_THRESHOLD;
+
+    if (tc->def->flags & TRAKTOR_MK2)
+        tc->threshold = ZERO_THRESHOLD_MK2;
+    else
+        tc->threshold = ZERO_THRESHOLD;
```
the xwax library has its own conventions which are more kernel-like.  It's better to match that style than to start to introduce mixxx style into those files.

jclsn replied:
I would also like to keep it this way, or better: Add it to the `struct timecode_def`

ywwg commented:
Please do not make style nits -- we are trying to quickly get these changes in for quick release, and I'd prefer to stay focused on functionality. @jclsn is correctly hewing to the xwax style for those changes, even though it doesn't match our in-house style.

jclsn commented on src/vinylcontrol/defs_vinylcontrol.h:
```diff
@@ -53,6 +53,9 @@ constexpr int VINYL_STATUS_ERROR = 3;
 
+#define MIXXX_VINYL_PITCH_FILTER_LEGACY "Alpha-Beta filter (fast)"
+#define MIXXX_VINYL_PITCH_FILTER_KALMAN "Kalman filter (accurate)"
+
```
Should I change this now or not?

Swiftb0y replied:
This insinuates to me that I would want the direct control of the Alpha-Beta filter, which we concluded is almost always the wrong choice, right? I mean, we flatout wanted to replace the alpha-beta filter because we figured it was universally superior, so if we keep the options, we should at least make it clear which one is the better default.

jclsn replied:
Alright, keeping it as is

Swiftb0y commented on lib/xwax/timecoder.c:
```diff
@@ -615,6 +639,59 @@ static void process_bitstream(struct timecoder *tc, signed int m)
+    tc->last_quadrant = tc->quadrant;
+    tc->direction_changed = direction_changed;
+
+    if (tc->primary.swapped && tc->primary.positive)
+        tc->quadrant = 1;
+    else if (tc->primary.swapped && !tc->primary.positive)
+        tc->quadrant = 3;
+    else if (tc->secondary.swapped && tc->secondary.positive)
+        tc->quadrant = 2;
+    else if (tc->secondary.swapped && !tc->secondary.positive)
+        tc->quadrant = 4;
```
That was exactly my question :sweat_smile: I mean since the math basically always just looks at the difference (derivate) and is $$\mod 4$$ anyways, it doesn't really make a difference at all. In that case I'd prefer if we stay in the 0-based domain. 

This also gives rise to the consequence that the `tc->secondary.swapped && !tc->secondary.positive` case is the same as `!tc->primary.swapped && !tc->secondary.swapped`. Is that conceptually correct?

jclsn replied:
No, it's not the same. The swapping means that one channel crossed the x-axis and changed polarity. Just because the primary channel did not swap, it doesn't mean that the secondary channel swapped. There are sample in  between those states.

I am not really sure what practical use 0,1,2,3 has over 1,2,3,4.

jclsn replied:
Changed

Swiftb0y commented on lib/xwax/.clang-format:
out of curiosity, where are these values from? Can this be shortened by basing it upon another style? Clang-format has this very concise example for kernel-like formatting for instance: https://clang.llvm.org/docs/ClangFormatStyleOptions.html#examples

jclsn committed:
xwax: Refine Kalman filter thresholds

I realized that the Traktor MK2 vinyls were way too sensitive and the
Seratos too unsensitive. The modes switched too hard, so I added a new
mode. The transitions are now a bit smoother. I think I have to rework
the model in the long run though.

jclsn commented:
So I just tried the new Traktor MK2 2024 Edition. It behaves quite differently, as it's a bit heavier. Maybe what I was experiencing was due to the older record just being slippier. I am starting to think if I should just let be for now and maybe do a dedicated testing session with all vinyl types.

ywwg commented on lib/xwax/pitch_kalman.h:
```diff
@@ -58,23 +59,24 @@ struct pitch_kalman {
 
     /* Tresholds of the innovation quantity for the mode switches */
```
hmmm still not quite sure what is meant by innovation again

```suggestion
    /* Thresholds of the innovation quantity for the mode switches */
```

ywwg commented on lib/xwax/timecoder.c:
```diff
@@ -499,6 +516,57 @@ static void process_bitstream(struct timecoder *tc, signed int m)
+    static unsigned long long direction_change_counter = 0;
+
+    /* Check for a displacement of four quadrants */
+    if (diff == 0 && !tc->direction_changed) {
+        return 1.0 / tc->def->resolution;
+
+    /* Check for a displacement of three quadrants */
+    } else if ((tc->forwards && diff == 3) || (!tc->forwards && diff == 1)) {
+        return (3.0 / tc->def->resolution) / 4.0;
+
+    /* Check for a displacement of two quadrants  */
+    } else if (diff == 2) {
+        return (1.0 / tc->def->resolution) / 2.0;
+
+    /* Fallback */
+    } else {
+        return (1.0 / tc->def->resolution) / 4.0;
+    }
```
with returns, no need for `else`, it's implied.  and "fallback" is intuitive from the code flow, so we don't need that comment.  Only making this nit because it's easy to produce a suggestion.

```suggestion
    if (diff == 0 && !tc->direction_changed) {
        return 1.0 / tc->def->resolution;
    }
    
    /* Check for a displacement of three quadrants */
    if ((tc->forwards && diff == 3) || (!tc->forwards && diff == 1)) {
        return (3.0 / tc->def->resolution) / 4.0;
    }
    
    /* Check for a displacement of two quadrants  */
    if (diff == 2) {
        return (1.0 / tc->def->resolution) / 2.0;
    }
    
    return (1.0 / tc->def->resolution) / 4.0;
```

ywwg reviewed (commented):
one language question and one nit.  I will try this at home with serato, traktor 1 and traktor 2 vinyl later.

Swiftb0y commented on lib/xwax/pitch_kalman.c:
```diff
@@ -167,17 +183,26 @@ void pitch_kalman_update(struct pitch_kalman* p, double dx)
     const double y_abs = fabs(y);
 
-    if (y_abs > p->reactive_threshold)
+    kalman_debug("innovation: %+f, ", y);
```
yeah, its standard terminology from what I've gathered.

ywwg commented on lib/xwax/timecoder.c:
```diff
@@ -31,8 +31,6 @@
 #include "debug.h"
 #include "timecoder.h"
 
-#define ZERO_THRESHOLD (128 << 16)
```
why did we drop this definition? we repeatedly use this value in the structs so it seemed useful?

Swiftb0y replied:
The traktor mk2 needed a different threshold from Jan's experiments. Which is why he replaced the global with the per-struct definition. Those changes to the mk2 definitions got lost during the rebase. I guess he removed the `#define` since there was no use to have multiple different global `ZERO_TRESHOLD` definitions?

ywwg commented:
unfortunately this seems to break Traktor MK1 vinyl. It stutters during the leadin and then goes way too fast in the main track.  this pr: 

https://github.com/user-attachments/assets/9f630a93-eb86-49d1-b6ef-58f88191c12f

ywwg commented:
I will see if it was working at all before the work we've been doing, or with the other filters.

ywwg reviewed (approved):
Can't test MK1 right now but it's not blocking. We'll address that next.
